### ISSUE TITLE
consistent type signatures

### ISSUE THREAD

**Issue by davidchambers** [ramda/ramda#1554]
Questions to be answered:

  - __How should pairs be represented?__ `[a, b]` and `(a, b)` are leading contenders.

  - __How should binary functions be represented?__ `(a, b) -> c` is one option, though this may conflict with the representation of pairs.

  - __How should "dictionaries" be represented?__ `{a}`, `{k: v}`, and `StrMap a` are possibilities.

  - __How should "records" be represented?__ `{ age :: Number, name :: String }` makes sense when the fields are known, but `Object` might be all we can say for functions such as `R.prop` which operate on any record.

**Comment by scott-christopher** [comment]
My thoughts:
* **How should pairs be represented?**
   * My preference would be for `[a, b]`, as it better reflects that it just represents a two element array. The downside however would be that there could be some confusion in thinking `[a]` then represents a single element array, rather than a list of any size. We could consider something like `[...a]` to represent lists of any size.

* **How should binary functions be represented?**
   * If the binary function should be curried, then `a -> b -> c`, however I'd opt for `(a, b) -> c` in other cases assuming we went with `[a, b]` for pairs. 

* **How should "dictionaries" be represented?**
   * My preference is probably `{a}` here, as `k` is always going to be `String` for native JS objects and the shorthand is nicer in my opinion.

* **How should "records" be represented?**
   * I agree that the keys should be present if the keys help to define the type. I'm somewhat inclined to think we should just represent "records" in the same way as "dictionaries", as `{a}` allows for `a` to be referenced elsewhere in the type signature, even if it just represents a type in the set of all types used for values in the object. For example what would the return type of `prop` look like if `Object` were used in the type signature rather than `{a}`?

**Comment by svozza** [comment]
If we're worried about confusion arising from using brackets for pairs and functions then we might want to do something about the two different ways they're used in the type sig of the compose* functions.

**Comment by CrossEye** [comment]
Another one I forgot to respond to.  (This one I actually saw when it first came in, too.  Shame on me!)

* **How should pairs be represented?**
   * While I prefer `[a, b]`, over `(a, b)`, I do wonder if we might not be still better off by considering `<a, b>`, which wouldn't allow the confusion with `[a]`.

* **How should binary functions be represented?**
   * I agree with @scott-christopher: `a -> b -> c` for curried function, `(a, b) -> c` for non-curried ones.

* **How should "dictionaries" be represented?**
   * Here I _think_ I agree again with @scott-christopher that `{a}` is best here.  But I'm less convinced.  Nothing seems quite right.

* **How should "records" be represented?**
   * I like @davidchambers' notion of `{ age :: Number, name :: String }` for known fields, although I'm not sure how often that will affect Ramda's actual API. When they're not known, then  `{a}` makes sense.

I believe there are several other important questions.

* **How do we represent variadic functions?**  Although Ramda itself has been moving away from having variadic functions in its API, some of the functions supplied to it or returned by it might still be variadic.  We've dealt with these in an entirely _ad hoc_ fashion.  (See for instance, the difference between [`flip`][fl] and [`either`][ei].)  Can we create a consistent set of rules?

* **How and where should we document any types used as constraints?** Right now I can think of only `Ord` and `Lens`, but there may be others.  Our documentation should supply some sort of links to descriptions of them.  How do we deal with that?

* **Should we introduce an `Any` rather than `*` for arbitrary unknown types?**  The use of `a`, `b`, etc, is usually reserved for cases when the type is reused.  But the asterisk is somewhat obscure and even somewhat hard to see.


  [ei]: http://ramdajs.com/0.19.0/docs/#either
  [fl]: http://ramdajs.com/0.19.0/docs/#flip

**Comment by davidchambers** [comment]
> __Should we introduce an `Any` rather than `*` for arbitrary unknown types?__

:thumbsup:

**Comment by svozza** [comment]
Yeah, I like that idea too. How often should it be used? I can see a lot of type sigs having to change if they're to be strictly accurate. For example,

```concat: [a] -> [a] -> [a]``` should really be ```[Any] -> [Any] -> [Any]``` given this:

```js
R.concat([1, 'a', true], [{a:1}, 6.7, [2, 3]])
// => [1, "a", true, {"a": 1}, 6.7, [2, 3]]
```
You could say the same for many of the `List` functions.

**Comment by stoeffel** [comment]
@svozza sidenote: 
> concat: [a] -> [a] -> [a] should really be [Any] -> [Any] -> [Any] given this:

This is actually not fully right, since concat also works with strings and numbers:
```js
R.concat('The answer is: ', 42)
// => "The answer is: 42"
```

**Comment by svozza** [comment]
Heh. Down the rabbit hole of JavaScript's type system...

**Comment by CrossEye** [comment]
> I can see a lot of type sigs having to change if they're to be strictly accurate. For example,
> 
> ```concat: [a] -> [a] -> [a]``` should really be ```[Any] -> [Any] -> [Any]``` 

I would rather have aspirational type annotations than technically accurate ones.  If we mean the function to be used as `[a] -> [a] -> [a]`, then we should document it as such.  If we document how it's meant to be used, then we can feel free to change the implementation in a manner that breaks compatibility with the accidents of the current implementation so long as it still matches the desired behavior.  If we document every quirk of the current implementation, then any change, even an improvement in meeting what we _want_ it to do is a breaking change.

Also, there is no reason to -- and no hope in even trying to -- document all the signatures made possible by dispatching.  So `concat('abc', 'def'); //=> 'abcdef'` shouldn't be part of the discussion.

**Comment by svozza** [comment]
> If we document how it's meant to be used, then we can feel free to change the implementation in a manner that breaks compatibility with the accidents of the current implementation so long as it still matches the desired behavior. 

That seems like a perfectly sensible position.
